### ISSUE TITLE
Error in cime_cpl_init while running in singularity container

### ISSUE THREAD

**Issue by serbinsh** [ESMCI/cime#2929]
I am running into this error when attempting to execute a model after compilation 

```
 ERROR: (cime_cpl_init) :: namelist read returns an end of file or end of record condition
#0  0x7EFE64C92697
#1  0xADC5C6 in __shr_abort_mod_MOD_shr_abort_backtrace
#2  0xADC7E1 in __shr_abort_mod_MOD_shr_abort_abort
#3  0x41BC8F in __cime_comp_mod_MOD_cime_pre_init1
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1001.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
```

I am attempting to run CLM5 based on this testcase
./create_newcase --case testcase20 --compset I2000Clm50BgcCru --mach modex --res f09_f09 --project testing --run-unsupported

I am running in a singularity container with:
netCDF 4.6.2
HDF5 1.10.4
gcc 5.5.0
openMPI/mpirun 2.1.5

As i mentioned, I was able to build everything successfully but am hitting this error upon ./case.submit

**Comment by serbinsh** [comment]
Please let me know if this should be posted elsewhere but it seemed like a CIME issue, though could be CTSM?

**Comment by ekluzek** [comment]
This looks like an error in reading the namelist in cime_cpl_init. Look at the driver namelist carefully and see what might be wrong with it. The traceback should also give you the traceback so you can figure out which namelist it is. You might also try making some changes in the namelist to get past the problem...

**Comment by serbinsh** [comment]
Thanks @ekluzek Real quick, which is the "driver" namelist that cime_cpl_init is trying to access?  Also not sure I know how to run a traceback on the error stack or which log it is in.  I cant seem to find the detailed log in cesm log.  I guess I am not exactly sure which nl it is complaining about

**Comment by ekluzek** [comment]
I was hoping it would be in the cesm.log file. It did seem to do a traceback so I expected that it would go there. You can check the driver code for the write statement it does when it dies and see if that helps you figure out the namelist. But, it also might have the same error for more than one namelist. Hence, why I was hoping it would be in the cesm.log file. But, you might also clean the build and set DEBUG=TRUE and try it again.

**Comment by serbinsh** [comment]
Trying debug now....hope it will provide a bit more.

**Comment by serbinsh** [comment]
A little more output

```
#0  0x7FCCD32A4697
#1  0x126B604 in __shr_abort_mod_MOD_shr_abort_backtrace at shr_abort_mod.F90:104
#2  0x126B6E4 in __shr_abort_mod_MOD_shr_abort_abort at shr_abort_mod.F90:61
#3  0x4079D9 in __cime_comp_mod_MOD_cime_cpl_init at cime_comp_mod.F90:4105
#4  0x41D578 in __cime_comp_mod_MOD_cime_pre_init1 at cime_comp_mod.F90:607
#5  0x420A44 in MAIN__ at cime_driver.F90:37
```

**Comment by ekluzek** [comment]
So look on line 4105 of cime_comp_mod.F90 that's in cime/src/drivers/mct/main.

**Comment by serbinsh** [comment]
@ekluzek so here is the block of code containing 4105

```
  if (mype == 0) then
    ! Read coupler namelist if it exists
    ninst_driver = 1
    nu = shr_file_getUnit()
    open(unit = nu, file = NLFileName, status = 'old', iostat = ierr)
    rewind(unit = nu)
    ierr = 1
    do while ( ierr /= 0 )
       read(unit = nu, nml = cime_driver_inst, iostat = ierr)
       if (ierr < 0) then
          call shr_sys_abort( subname//':: namelist read returns an'// &
               ' end of file or end of record condition' )
       endif
    enddo
    close(unit = nu)
    call shr_file_freeUnit(nu)
    num_inst_driver = max(ninst_driver, 1)
  end if
```

Specifically this is 4105 ' end of file or end of record condition' ) or what we see in the error log.  

I am still trying to figure out what namelist it is reading here since perhaps its actually not being generated or in the case dir? Perhaps I am not setting up the case properly?

**Comment by jedwards4b** [comment]
It's reading the namelist cime_driver_inst from the file drv_in that namelist has exactly one variable 
```
&cime_driver_inst
  ninst_driver = 1
/
```

**Comment by serbinsh** [comment]
Is this the nl its trying to use?

&modelio
  diri = "/root/ctsm_output/CLM5_1543595736/bld/cpl"
  diro = "/root/ctsm_output/CLM5_1543595736/run"
  logfile = "cpl.log.181130-114530"
/
&pio_inparm
  pio_netcdf_format = "64bit_offset"
  pio_numiotasks = -99
  pio_rearranger = 1
  pio_root = 1
  pio_stride = 1
  pio_typename = "netcdf"
/

**Comment by jedwards4b** [comment]
@serbinsh The file is drv_in, the namelist is cime_driver_inst.

**Comment by serbinsh** [comment]
@jedwards4b thanks.  Indeed I see that list and it is setup as you posted.  not sure why it is complaining about this list then.

**Comment by serbinsh** [comment]
@jedwards4b I changed the dir permissions all to 777 and it still crashes.  I will try the same run again on our modex cluster but it seems it runs fine there.  I thought perhaps it had to do with the way the exe was being run with the MPI libs but the run crashes even if I  just run /root/ctsm_output/CLM5_1543597800/bld/cesm.exe

that said I have mostly been testing using --no-batch

**Comment by jedwards4b** [comment]
You might try replacing 
```
<entry id="run_exe" value="${EXEROOT}/cesm.exe ">
```
in env_mach_specific.xml with
```
<entry id="run_exe" value="cat drv_in ">
```
to see if that sheds any light on the problem.

**Comment by serbinsh** [comment]
is there somewhere I can look at details what specific programs or code is used to read namelists in the compiled binary?  Perhaps there is something related to the libs or versions I used?  I can run the same cases on my modex cluster, I just cannot seem to get past this namelist error on my singularity build...but its a pretty basic build and mimics my cluster setup

**Comment by serbinsh** [comment]
@jedwards4b sorry for some reason I didnt see your response until after i posted my message, I will give it a try

**Comment by serbinsh** [comment]
Here is what the cesm log shows:

```
less cesm.log.181130-165449

&ccsm_pes
  atm_layout = "concurrent"
  atm_ntasks = 1
  atm_nthreads = 1
  atm_pestride = 1
  atm_rootpe = 0
  cpl_ntasks = 1
  cpl_nthreads = 1
  cpl_pestride = 1
  cpl_rootpe = 0
  esp_layout = "concurrent"
  esp_ntasks = 1
  esp_nthreads = 1
  esp_pestride = 1
  esp_rootpe = 0
  glc_layout = "concurrent"
  glc_ntasks = 1
  glc_nthreads = 1
  glc_pestride = 1
  glc_rootpe = 0
  ice_layout = "concurrent"
  ice_ntasks = 1
  ice_nthreads = 1
  ice_pestride = 1
  ice_rootpe = 0
  lnd_layout = "concurrent"
  lnd_ntasks = 1
  lnd_nthreads = 1
  lnd_pestride = 1
  lnd_rootpe = 0
  ocn_layout = "concurrent"
  ocn_ntasks = 1
  ocn_nthreads = 1
  ocn_pestride = 1
  ocn_rootpe = 0
  rof_layout = "concurrent"
  rof_ntasks = 1
  rof_nthreads = 1
  rof_pestride = 1
  rof_rootpe = 0
  wav_layout = "concurrent"
  wav_ntasks = 1
  wav_nthreads = 1
...

Basically shows the cat output of the file

**Comment by serbinsh** [comment]
Hang on....I think I found out what was wrong, though I have no idea why this would have caused the problem.

I am still trying to sort my way around using a singularity container for CTSM and presently I am dealing with building a test environment in the shell.  But when you use the shell you are "root" which leads to other challenges. For example you shouldnt really run mpirun as root so a work around is to use ```--allow-run-as-root``` 

I was trying to just run the mpirun command myself, for example

```
mpirun -np 1 -npernode 1 --allow-run-as-root /root/ctsm_output/CLM5_1543611563/bld/cesm.exe
```

or just running the exe like

```
/root/ctsm_output/CLM5_1543597800/bld/cesm.exe
```

and I would get the error but I just tried this

```
  <mpirun mpilib="default">
    <executable>mpirun</executable>
    <arguments>
      <arg name="num_tasks">-np $TOTALPES</arg>
      <arg name="tasks_per_node">-npernode $MAX_TASKS_PER_NODE --allow-run-as-root</arg>
    </arguments>
  </mpirun>
```

then tried ./case.submit, and, it seems, that worked.....

```
Finished creating component namelists
Checking that inputdata is available as part of case submission
Loading input file list: 'Buildconf/mosart.input_data_list'
Loading input file list: 'Buildconf/datm.input_data_list'
Loading input file list: 'Buildconf/clm.input_data_list'
Model clm no file specified for finidat
Loading input file list: 'Buildconf/cpl.input_data_list'
Check case OK
submit_jobs case.run
job is case.run
Submit job case.run
Starting job script case.run
Creating component namelists
   Calling /root/ctsm/cime/src/components/data_comps/datm/cime_config/buildnml
   Calling /root/ctsm/cime_config/buildnml
WARNING: CLM is starting up from a cold state
   Calling /root/ctsm/cime/src/components/stub_comps/sice/cime_config/buildnml
   Calling /root/ctsm/cime/src/components/stub_comps/socn/cime_config/buildnml
   Calling /root/ctsm/components/mosart//cime_config/buildnml
   Calling /root/ctsm/cime/src/components/stub_comps/sglc/cime_config/buildnml
   Calling /root/ctsm/cime/src/components/stub_comps/swav/cime_config/buildnml
   Calling /root/ctsm/cime/src/components/stub_comps/sesp/cime_config/buildnml
   Calling /root/ctsm/cime/src/drivers/mct/cime_config/buildnml
Finished creating component namelists
-------------------------------------------------------------------------
 - Prestage required restarts into /root/ctsm_output/CLM5_1543597800/run
 - Case input data directory (DIN_LOC_ROOT) is /data/Model_Data/cesm_input_datasets
 - Checking for required input datasets in DIN_LOC_ROOT
-------------------------------------------------------------------------
2018-11-30 17:05:17 MODEL EXECUTION BEGINS HERE
run command is mpirun -np 1 -npernode 1 --allow-run-as-root /root/ctsm_output/CLM5_1543597800/bld/cesm.exe   >> cesm.log.$LID 2>&1
```

Huh?  I know I have just run the exe before without this problem, at least when testing CLM-FATES.

**Comment by serbinsh** [comment]
...that said, the ultimate solute is to setup the container to run externally and avoid a root user.  I suppose this was an issue from the circumstance of attempting to start a run inside the container

**Comment by rljacob** [comment]
I think you are blazing a new trail in trying to get CESM to run in a singularity container.

I would start with just trying to get CIME standalone to run.  Clone it from this repo and build a cime-only case like --compset X --res f19_g16

**Comment by serbinsh** [comment]
![screen shot 2018-11-30 at 5 11 04 pm](https://user-images.githubusercontent.com/3067369/49317601-20269680-f4c3-11e8-8f14-780ee62cd32c.png)

so its running, but as ...[clears throat].... as root.  Good thing I have a test cluster for these things.

**Comment by thiagoveloso** [comment]
I just wanted to report that I'm also experiencing this error. However, my model conditions are slightly different: I'm running a single-point CLM5 case (https://escomp.github.io/ctsm-docs/doc/build/html/users_guide/running-PTCLM/introduction-to-ptclm.html) on my MacPro machine. I am using the GNU compilers v.8.2.0, NetCDF-C v4.6.1 and NetCDF-fortran v6.4.4.

```
./create_newcase --user-mods-dir $MYDATAFILES/1x1pt_$SITE --case $CASEROOT --res CLM_USRDAT --compset ${COMPSET}
```

After the preparation of all surface files required by the case, I get to the point where everything runs smoothly until `./case.submit`. Once there, I get:

```
Creating component namelists
   Calling /Users/thiago/models/clm5.0/cime/src/components/data_comps/datm/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/sice/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/socn/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/components/mosart//cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/sglc/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/swav/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/sesp/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/drivers/mct/cime_config/buildnml
Finished creating component namelists
Checking that inputdata is available as part of case submission
Loading input file list: 'Buildconf/cpl.input_data_list'
Loading input file list: 'Buildconf/datm.input_data_list'
Loading input file list: 'Buildconf/mosart.input_data_list'
Loading input file list: 'Buildconf/clm.input_data_list'
Check case OK
submit_jobs case.run
Submit job case.run
Starting job script case.run
Generating namelists for /Users/thiago/cesm-cases/testPTCLM
Creating component namelists
   Calling /Users/thiago/models/clm5.0/cime/src/components/data_comps/datm/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/sice/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/socn/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/components/mosart//cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/sglc/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/swav/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/components/stub_comps/sesp/cime_config/buildnml
   Calling /Users/thiago/models/clm5.0/cime/src/drivers/mct/cime_config/buildnml
Finished creating component namelists
-------------------------------------------------------------------------
 - Prestage required restarts into /Volumes/Storage/data/CLM/scratch/testPTCLM/run
 - Case input data directory (DIN_LOC_ROOT) is /Volumes/Storage/data/CLM/ptclm_data_files
 - Checking for required input datasets in DIN_LOC_ROOT
-------------------------------------------------------------------------
2019-01-07 20:15:05 MODEL EXECUTION BEGINS HERE
run command is   /Volumes/Storage/data/CLM/scratch/testPTCLM/bld/cesm.exe  >> cesm.log.$LID 2>&1
ERROR: RUN FAIL: Command '  /Volumes/Storage/data/CLM/scratch/testPTCLM/bld/cesm.exe  >> cesm.log.$LID 2>&1 ' failed
See log file for details: /Volumes/Storage/data/CLM/scratch/testPTCLM/run/cesm.log.190107-201504
```
I have also attached the log files, which unfortunately don't reveal any further information.

When I try invoking the executable manually, i.e. `/Volumes/Storage/data/CLM/scratch/testPTCLM/bld/cesm.exe` , this is what I get:

```
 ERROR: (cime_cpl_init) :: namelist read returns an end of file or end of record condition
#0  0x109f0cf8f
#1  0x103c6572f
#2  0x103c658db
#3  0x1036cf539
#4  0x103e3e11c
MPI_Abort: error code = 1001
```
Running the executable as root, like @serbinsh did, doesn't work for me. 

Any suggestions on how to further debug this error?

P.S.: Although I'm not using Singularity, I decided not to open a new issue because there's already such a rich discussion on this error.

[atm.log.190107-204122.txt](https://github.com/ESMCI/cime/files/2734926/atm.log.190107-204122.txt)
[cesm.log.190107-204122.txt](https://github.com/ESMCI/cime/files/2734927/cesm.log.190107-204122.txt)
[cpl.log.190107-204122.txt](https://github.com/ESMCI/cime/files/2734928/cpl.log.190107-204122.txt)

**Comment by jedwards4b** [comment]
If you want to try to execute the system manually you need to do so from the run directory.

The error that you got seems to indicate that you've somehow confused serial and parallel builds.
You might try creating the case again with the addition of flag ```--mpilib mpi-serial``` 

Finally this is not the correct forum for this discussion - please use  http://forum.cgd.ucar.edu/
This forum is appropriate if you have clearly identified a bug in cime.

**Comment by thiagoveloso** [comment]
I ended up finding out what was wrong @jedwards4b.

When running this single-point experiment, I was using the same Hdf5 and Netcdf libraries compiled with parallel support that I used to run a regular global/regional experiment.

However, it worked fine when I built Hdf5 and Netcdf **without** the '--enable-parallel' and '--enable-parallel-tests' switches (respectively for Hdf5 and Netcdf). In my tests, this CLM5 single-point simulation (US-UMB) worked in both Linux and Mac machines with the regular (i.e. non-parallel) libraries.

This leads me to believe that for some reason the single-point mode is not compatible with the parallelized version of the required libraries.

I would highly recommend you guys to add some kind of note or warning about that on the instructions page at https://escomp.github.io/ctsm-docs/doc/build/html/users_guide/running-single-points/running-single-point-configurations.html.
